src/broadcast/broadcast.rs

```diff
@@ -250,14 +279,66 @@ impl<N: NodeIdT> Broadcast<N> {
         // Save the proof for reconstructing the tree later.
         self.echos.insert(sender_id.clone(), p);
 
-        if self.ready_sent || self.count_echos(&hash) < self.netinfo.num_correct() {
+        // Upon receiving `N - 2f` `Echo`s with this root hash, multicast `CanDecode`
+        if !self.can_decode_sent && self.count_echos(&hash) >= self.netinfo.num_correct() - self.netinfo.num_faulty() {
+            return self.send_can_decode(&hash);
+        }
+
+        if self.ready_sent || self.count_echos(&hash) + self.count_echo_hashes(&hash) < self.netinfo.num_correct() {
             return self.compute_output(&hash);
         }
 
         // Upon receiving `N - f` `Echo`s with this root hash, multicast `Ready`.
         self.send_ready(&hash)
     }
 
+    fn handle_echo_hash(&mut self, sender_id: &N, hash: &Digest) -> Result<Step<N>> {
+        // If the sender has already sent `EchoHash`, ignore.
+        if let Some(old_hash) = self.echo_hashes.get(sender_id) {
+            if old_hash == hash {
+                warn!(
+                    "Node {:?} received EchoHash({:?}) multiple times from {:?}.",
+                    self.our_id(),
+                    hash,
+                    sender_id,
+                );
+                return Ok(Step::default());
+            } else {
+                return Ok(Fault::new(sender_id.clone(), FaultKind::MultipleEchoHashes).into());
+            }
+        }
+        // Save the hash for counting later.
+        self.echo_hashes.insert(sender_id.clone(), *hash);
+
+        if self.ready_sent || self.count_echos(&hash) + self.count_echo_hashes(&hash) < self.netinfo.num_correct() {
```

Comment:
We can't just take this sum (sorry, I formulated this the wrong way in the issue!), because then we'd count a node that has sent us both an `Echo` and `EchoHash` twice!
I wonder whether it would be easier to just count every `Echo` as an `EchoHash` implicitly, too, i.e. also insert their hash into `echo_hashes`?

Reply:
We could take the length of the set union of `echo` and `echo_hash` keys where the value is `hash`. Counting every `Echo` as an `EchoHash` implicitly and then counting just the `EchoHash`s would work too but I think it would lead to some confusing code.

Reply:
That's true; but let's avoid actually allocating the set, and just count the number of node IDs that are present in either.
Alternatively, we could replace the two maps with one:
```rust
echo_hashes: BTreeMap<N, (Digest, Option<Proof<Vec<u8>>>)>,
```

Reply:
Single map idea is awesome. Thanks!